stream/message.go

```diff
@@ -0,0 +1,11 @@
+package stream
+
+import "time"
+
+// Message represents a single object received or sent to a stream, using one of
+// the configured stream consumers or producers.
+type Message struct {
```

Comment:
I know you want to abstract the underlying technology but for error reporting it would be very nice if we had partition + offset of the message. That way we don't need to have x error report topics in for streaming mailer. Just partition and offset would be enough, we can very easily pick that message out with kafkacat.

These fields could be private struct fields but with some getter methods so they could only be consumed and not set on a message

Reply:
Good one 👍 I also haven't fully fleshed out the message part yet. The "old" way was to have all of the implementation-specific values to be part of the struct, like here:

https://github.com/blendle/go-streamprocessor/blob/035a59eef797da175fd5aafdd152d9f5b8fcafea/stream/message.go#L11-L18

The downside is that you have nonsensical values for different implementations (like partition for Pubsub), the upside is that it's the most easy way to tackle this.

If you have any ideas on this, feel free to share them. If not, we'll go with your suggestion for now, and just accept that not all values are relevant based on the implementation you use.

Reply:
Thinking a bit on this, it might be useful to actually add client-specific details in separate structs, the same way I've proposed to doing in the client configurations:

https://github.com/blendle/go-streamprocessor/blob/34dc3d1790e845b23185a2f560412c70154865f0/streamconfig/config.go#L10-L20

So you'd get something like:

```golang
type Message struct {
  Value []byte
  Timestamp []byte

  Kafka *kafkamessage.Message struct {
    Key []byte
    Topic string
    ...
  }
  Pubsub *pubsubmessage.Message struct {
    ...
  }
}
```

It would make it a bit more "annoying" to read often-used primitives like the key in Kafka for example, but it does make for a more clean separation.

Reply:
Thinking some more on this. The above suggestion wouldn't be ideal (or at least needs some more thinking), since you don't want to panic whenever a stream processor for example uses `stream.Message.Kafka.Key`, but then someone starts that stream processor using the `standardstream` implementation by piping data to it. What to do at this point?

I'd at least like to _try_ and see if we can end up with a clean interface that allows this cross-implementation usage up to a certain point, without having to modify the code. Obviously, once you get into very specific usages, this no longer works, but maybe we can at least make sure that if you set `stream.Message.Kafka.Key`, and use the `standardstream` implementation, that struct value becomes a no-op.

Reply:
I'm not sure if having on interface for both Kafka and PubSub makes sense, they have very different characteristics and use cases, and I think that by having one interface that covers only the intersection of the different systems we might end up not being able to take full advantage of this.


I'll try to think about this more in-depth later

Reply:
> I'm not sure if having on interface for both Kafka and PubSub makes sense, they have very different characteristics and use cases

I've heard this before, but I don't actually agree with it, so far. Yes, there are differences, but I have a Pubsub implementation running locally (using the old code), and it fits perfectly well within the scope of what is currently available in the master branch. Like I said, I agree that there are _some_ differences (and we need to tackle those), but in general, you point both to read messages from somewhere, and you can publish your messages somewhere.

The biggest differences are actually in how the topics (or subscriptions for Pub/Sub) are configured, but since this package doesn't concern itself with creating and managing topics/subscriptions, that can still be configured individually, allowing your more fine-grained control.

Could you elaborate a bit more on what those differences are that we would want to leverage, that can't fit into a single interface that serves both implementations?

Remember: this discussion is about the _message_ interface/struct, not so much the configuration of the client, subscriber, or publisher, since those _will_ have separate tunables to fine-tune implementation-specific details, using the [`streamconfig` package](https://github.com/blendle/go-streamprocessor/pull/38).

Reply:
So, thinking some more on this. I think we can get _almost_ best of both worlds, by using the proper interfaces, and some type conversion.

Some considerations:

* Looking at some of the online benchmarks, I don't think the performance cost of using type conversion (or, type checking) is significant anymore, so I don't think we should worry about this at this stage.

* All clients will implement the most basic interface, which supports `Value() []byte`, `Timestamp() time.Time` and `Ack() error`.
* Then, each client will have their own interface that extends the basic with any additional functions, for example, Kafka will add (amongst others) `Key() []byte`.
* To work with this, `consumer.Message()` will return `interface{}`
* When implementing your own stream processor, you can decide what to do, based on the simplicity of the processor.

For example, this would be the most simple implementation:

```golang
for m := range consumer.Message() {
  msg := m.(streammsg.Message)
  fmt.Printf("received %s at unix: %d\n", string(msg.Value()), msg.Timestamp().Unix())
}
```

A more advanced processor that wants to do stuff based on the key sent by Kafka would implement:

```golang
for m := range consumer.Message() {
  if msg, ok := m.(kafkamsg.Message); ok {
    fmt.Printf("received %s with key %s\n", string(msg.Value()), string(msg.Key()))
  } else {
    panic()
  }
}
```

Or, if you want to support multiple use-cases:

```golang
for m := range consumer.Message() {
  switch msg := m.(type) {
  case kafkamsg.Message:
    fmt.Printf("received %s with key %s\n", string(msg.Value()), string(msg.Key()))
  case streammsg.Message:
    fmt.Printf("received %s without key\n", string(msg.Value()))
  default:
    panic()
  }
}
```

Some downsides:

* There's a bit more work involved to get things working (i.e. you need to do at least one type conversion)
* There's no guarantee that all stream processors built with this package actually work with all the implementations. (i.e. you can't just take a stream processor and have it work with Pubsub, without potentially first changing the code)
* There's a minor performance loss, but it's minor, and I don't consider it a downside that would convince me not to do this

Some more thoughts:

* I'm considering what the "base" message interface should be. `Value()` and `Timestamp()` certainly make sense. I'm also considering if `Key()` makes sense, as right now Kafka is the dominant implementation we use, and in most stream processors, we actually use the Key value

* I haven't thought through the producer implementation yet, but I'm guessing you'd create a message struct, either the basic one if that suffices, or an implementation specific one if you (for example) need to set the topic for a Kafka message.

* There's also the possibility that we want multiple message types that are implementation agnostic. I'm thinking along the lines of the [io package](https://golang.org/src/io/io.go?s=5240:5298#L128), which for example has interfaces such as `WriteCloser`, `ReadWriteCloser` and `ReadWriteSeeker`. I could see a possibility of using `KeyedMessage`, and `Message` for messages with and without a key.

* I'm not sure if the above implementation is the best solution, but it's _one_ potential solution. Let me know what you think or if you have any other proposals. In the mean time, I'll continue experimenting and playing around with this.

Reply:
One more quick addition. In fact, what we should do, instead of returning `interface{}`, is to actually return the `streammsg.Message` interface for _all_ consumer implementations. That way, there's no type conversion involved, _unless_ you want to do more advanced stuff.

So the simplest implementation would be:

```golang
for msg := range consumer.Message() {
  fmt.Printf("received %s at unix: %d\n", string(msg.Value()), msg.Timestamp().Unix())
}
```

And a more advanced scenario could be:

```golang
for m := range consumer.Message() {
  keyedMsg := m.(kafkamsg.Message) // need panic check here
  fmt.Printf("received %s with key %s\n", string(msg.Value()), string(keyedMsg.Key()))
}
```

Reply:
Excuse my ramblings, just thinking out loud here.

I'm testing some different ways of tackling this. One is to have several interfaces for specific use-cases, and then implement those interfaces in different clients. I'm not going to paste all the code here, as I'd like to get a PR up, but I was thinking about the process of `ack`ing a message (which, right now, is done with `msg.Done()`).

At first, I considered an interface for this as well:

```golang
// Acker interface allows messages to be acknowledged. Each stream client has
// its own implementation to determine what it means to ack a message.
type Acker interface {
  Ack() error
}
```

Obviously, not all implementations have the concept of "acking" a message. So, my first thought would be not to have this be part of all clients. If that's the case, then in the stream processor itself, you could do something like this:

```golang
for msg := range consumer.Message()
  // do something with the message...

  if m, ok := msg.(Acker); ok {
    m.Ack()
  }
}
```

This way, in your implementation, you'd only ack if the client supports acking. 

But when testing this, it just felt awkward. Especially since this is a situation where there's no downside to have a no-op implementation on `Ack` for all clients. Doing it that way, we'd simply make `Ack()` become part of the standard interface, and simply make it a no-op when it has no function.

Alright, back to the drawing board 🎉

Reply:
Initial experiment/implementation: https://github.com/blendle/go-streamprocessor/pull/42